### PR TITLE
Extract and replace data in overlapping regions between tiles

### PR DESCRIPTION
**Pull request by IgorTatarnikov**
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/brainglobe/.github/blob/main/CONTRIBUTING.md).

Please fill out as much of this template as you can, but if you have any problems or questions, just leave a comment and we will help out :)

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
In order to correct for brightness differences, and interpolate over the stitches in the final fused images it's important to keep track of which tiles overlap and the coordinates of the overlap in both `local` (relative to the tile itself) and `global` (relative to the final fused image) coordinates.   

**What does this PR do?**
* **`brainglobe_stitch/tile.py`**: Introduced the `Overlap` class to store and manage information about the overlap between two tiles, including methods to calculate local overlap coordinates and extract/replace overlap data.

* **`brainglobe_stitch/image_mosaic.py`**: Updated the `ImageMosaic` class to include an `overlaps` attribute and methods to calculate overlaps between tiles. This involved modifying the constructor, `load_mesospim_directory`, and `stitch` methods, and adding the `calculate_overlaps` method.

* **`tests/test_unit/test_tile.py`**: Added new tests for the `Overlap` class, including tests for initialization, local coordinate calculation, overlap data extraction, and data replacement. These tests ensure that the overlap functionality works correctly.

## How has this PR been tested?
Tests added to cover the new functionality.

NOTE:
To test the functionality locally you must install Fiji and also add BigStitcher to the update sites. This can be done using the GUI [see here](https://imagej.net/plugins/bigstitcher/#download) or via the command line after installing Fiji.

./ImageJ-linux64 --update add-update-sites "BigStitcherUpdate" "https://sites.imagej.net/BigStitcher/"
./ImageJ-linux64 --update update

Replace ImageJ-linux64 with the path to the local ImageJ executable.

Three different datasets are available for testing:

- Tiny (~28 MB): this is mostly used for automated testing and doesn't work for visual inspection. It can be downloaded [here](https://gin.g-node.org/IgorTatarnikov/brainglobe-stitch-test/raw/master/brainglobe-stitch/brainglobe-stitch-test-data.zip).
- Small (~1.8 GB): this dataset is more useful for visual inspection. It can be downloaded [here](https://gin.g-node.org/IgorTatarnikov/brainglobe-stitch-test/raw/master/brainglobe-stitch/big_test_data.zip).
- A full size dataset is available (~130 GB): this dataset works best for testing overall performance, especially when reading or writing data. It's available by request.

## Is this a breaking change?
No

## Does this PR require an update to the documentation?
Yes, will come in subsequent PR.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality (unit & integration)
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: brainglobe_stitch/image_mosaic.py [hunk 1]
```diff
@@ -348,6 +355,48 @@ def read_big_stitcher_transforms(self) -> None:
...
+        for tile_i in self.tiles[:-1]:
+            position_i = tile_i.position
+            tile_i.neighbours = []
+            for tile_j in self.tiles[tile_i.id + 1 :]:
+                position_j = tile_j.position
+
+                # Check for overlap in the x and y dimensions
+                # and that the tiles do not have the same tile_id
+                # and that the tiles are from the same channel
+                if (
```
Comment (lauraporta) [inline brainglobe_stitch/image_mosaic.py:374]: Took me a while to parse these conditions 🤣

File: brainglobe_stitch/tile.py [hunk 2]
```diff
@@ -55,3 +55,148 @@ def __init__(
...
+        overlap_coordinates: npt.NDArray
+            The starting coordinates for the overlap in the fused image.
+        overlap_size: npt.NDArray
+            The size of the overlap in the fused image [z,y,x].
+        tile_i: Tile
+            The first tile.
+        tile_j: Tile
+            The second tile.
+        """
+        self.coordinates: npt.NDArray = overlap_coordinates
```
Comment (lauraporta) [inline brainglobe_stitch/tile.py:100]: Oh nice didn't know about `numpy.typing.NDArray`... I was just using `np.ndarray`

File: brainglobe_stitch/tile.py [hunk 3]
```diff
@@ -55,3 +55,148 @@ def __init__(
...
+        self.size: List[npt.NDArray] = [overlap_size]
+        self.tiles: Tuple[Tile, Tile] = (tile_i, tile_j)
+        self.local_coordinates: List[Tuple[npt.NDArray, npt.NDArray]] = [
```
Comment (lauraporta) [inline brainglobe_stitch/tile.py:103]: Is this datatype (`List[Tuple[npt.NDArray, npt.NDArray]]`) convenient for you somewhere else? 😮

File: brainglobe_stitch/tile.py [hunk 4]
```diff
@@ -55,3 +55,148 @@ def __init__(
...
+    def get_local_overlap_coordinates(self) -> None:
+        """
+        Calculate the starting coordinates for the overlap in
+        each tile's data pyramid.
+        """
+        tile_shape = self.tiles[0].data_pyramid[0].shape
+        resolution_pyramid = self.tiles[0].resolution_pyramid
+
+        for i in range(self.coordinates.shape[0]):
+            if self.tiles[0].position[i] < self.tiles[1].position[i]:
```
Comment (lauraporta) [inline brainglobe_stitch/tile.py:120]: Why are we referring to the first and second tile? 🤔

Comment (lauraporta) [reply brainglobe_stitch/tile.py:120]: I am understanding that the list of tiles is calculated once outside the creation of the `Overlap` object, and so they will have fixed positions... how do the first and second tile relate to the local coords of any overlap?

File: brainglobe_stitch/tile.py [hunk 5]
```diff
@@ -55,3 +55,148 @@ def __init__(
...
+        for j in range(1, len(resolution_pyramid)):
+            self.local_coordinates.append(
+                (
+                    self.local_coordinates[0][0] // resolution_pyramid[j],
+                    self.local_coordinates[0][1] // resolution_pyramid[j],
+                )
+            )
+            self.size.append(self.size[0] // resolution_pyramid[j])
+
+    def extract_tile_overlaps(
```
Comment (lauraporta) [inline brainglobe_stitch/tile.py:140]: Do you think that these two methods will be useful for the plugin as well (and not uniquely to testing)?

File: tests/test_unit/test_tile.py [hunk 6]
```diff
@@ -1,4 +1,106 @@
-from brainglobe_stitch.tile import Tile
+from typing import List, Tuple
```
Comment (lauraporta) [inline tests/test_unit/test_tile.py:1]: Don't know if you only use `tox` for testing, but with `pytest` the tests do not run. I think the `pyproject.toml` file needs a change:
```
[tool.pytest.ini_options]
addopts = ". --cov=brainglobe_stitch"
```

**Review by lauraporta — COMMENTED** [review]
Hi Igor, looks good to me. I'm just confused about how does `get_local_overlap_coordinates` work. Once I understand this bit, I'll complete the review with the tests.

File: brainglobe_stitch/tile.py (hunk 3 above)
Comment (lauraporta) [reply brainglobe_stitch/tile.py:103]: I later understood that the `tuple` level is for the resolution 🙂
Can we have a comment unpacking the hierarchies of `local_coordinates`?

File: brainglobe_stitch/tile.py (hunk 5 above)
Comment (IgorTatarnikov) [reply brainglobe_stitch/tile.py:140]: They'll be used in upcoming PRs for replacing overlapping areas after interpolating the stitched area.

File: tests/test_unit/test_tile.py (hunk 6 above)
Comment (IgorTatarnikov) [reply tests/test_unit/test_tile.py:1]: I can't seem to replicate this. What's the error?

File: brainglobe_stitch/tile.py (hunk 3 above)
Comment (IgorTatarnikov) [reply brainglobe_stitch/tile.py:103]: Can do!

File: brainglobe_stitch/tile.py (hunk 2 above)
Comment (IgorTatarnikov) [reply brainglobe_stitch/tile.py:100]: It's very handy! You can also specify the specific type the array is holding.

File: brainglobe_stitch/image_mosaic.py (hunk 1 above)
Comment (IgorTatarnikov) [reply brainglobe_stitch/image_mosaic.py:374]: They're a mouthful!

File: brainglobe_stitch/tile.py (hunk 4 above)
Comment (IgorTatarnikov) [reply brainglobe_stitch/tile.py:120]: Each `Overlap` is made up of two tiles. There are two separate but connected sets of coordinates that we keep track of, the "global" coordinates representing where the overlap lies in the final fused image, and the "local" coordinates which hold information on where the overlap lies in the `Tile` it originates in. The `get_local_overlap_coordinates` method checks which tile is "first" in each dimension to see if the overlap is at the beginning or the end for that specific dimension. This later allows us to extract, and modify the data stored in each overlap prior to fusing the entire image.

Comment (lauraporta) [reply brainglobe_stitch/tile.py]: Oh wow I would have never guessed that 😂

**Review by lauraporta — APPROVED** [review]
LeGiTiMate ✅

File: tests/test_unit/test_tile.py (hunk 6 above)
Comment (lauraporta) [reply tests/test_unit/test_tile.py:1]: No idea, now it works without any issue 🤷🏻‍♀️

**Pull request merged**